tools/unit/src/main/kotlin/com/datadog/tools/unit/extensions/TestConfigurationExtension.kt

```diff
@@ -0,0 +1,110 @@
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.tools.unit.extensions
+
+import android.annotation.SuppressLint
+import com.datadog.tools.unit.annotations.TestConfigurationsProvider
+import com.datadog.tools.unit.extensions.config.TestConfiguration
+import fr.xgouchet.elmyr.Forge
+import fr.xgouchet.elmyr.junit5.ForgeExtension
+import org.junit.jupiter.api.extension.AfterAllCallback
+import org.junit.jupiter.api.extension.AfterEachCallback
+import org.junit.jupiter.api.extension.BeforeAllCallback
+import org.junit.jupiter.api.extension.BeforeEachCallback
+import org.junit.jupiter.api.extension.ExtensionContext
+
+/**
+ * A JUnit Jupiter extension that can ensure a test configuration state is properly handled with
+ * self contained independent configurations.
+ */
+class TestConfigurationExtension :
+    BeforeAllCallback,
+    BeforeEachCallback,
+    AfterEachCallback,
+    AfterAllCallback {
+
+    private val testConfigurations: MutableMap<String, List<TestConfiguration>> = mutableMapOf()
+
+    // region BeforeAllCallback
+
+    /** @inheritdoc */
+    override fun beforeAll(context: ExtensionContext?) {
+        checkNotNull(context)
+        val methods = context.requiredTestClass
+            .methods
+            .filter {
+                it.isAnnotationPresent(TestConfigurationsProvider::class.java)
+            }
+
+        val configs = methods.flatMap { it.invoke(null) as List<*> }
+            .filterIsInstance<TestConfiguration>()
+        if (configs.isNotEmpty()) {
+            testConfigurations[context.uniqueId] = configs
+        }
+    }
+
+    // endregion
+
+    // region BeforeEachCallback
+
+    /** @inheritdoc */
+    override fun beforeEach(context: ExtensionContext?) {
+        context?.callTestConfigurations { forge -> setUp(forge) }
+    }
+
+    // endregion
+
+    // region AfterEachCallback
+
+    /** @inheritdoc */
+    override fun afterEach(context: ExtensionContext?) {
+        context?.callTestConfigurations { forge -> tearDown(forge) }
+    }
+
+    // endregion
+
+    // region AfterAllCallback
+
+    /** @inheritdoc */
+    override fun afterAll(context: ExtensionContext?) {
+        checkNotNull(context)
+        testConfigurations.remove(context.uniqueId)
+    }
+
+    // endregion
+
+    // region Internal
+
+    private fun ExtensionContext.callTestConfigurations(
+        operation: TestConfiguration.(Forge) -> Unit
+    ) {
+        val forge = getForge()
+        checkNotNull(forge) { "Unable to get the active Forge instance from the global store." }
+        callTestConfigurations(forge, operation)
+    }
+
+    @SuppressLint("NewApi")
+    private fun ExtensionContext.callTestConfigurations(
+        forge: Forge,
+        operation: TestConfiguration.(Forge) -> Unit
+    ) {
+        testConfigurations[uniqueId]?.forEach {
+            it.operation(forge)
+        }
+
+        if (parent.isPresent) {
+            parent.get().callTestConfigurations(forge, operation)
```

Comment:
I'm not 100% clear with that. I understand that test configuration `before` and `after` will wrap particular test execution like `BeforeEach` and `AfterEach`, so to follow the chain starting from `beforeEach` method of this class:

1. Initially nothing in this block will be executed, because there are no configurations registered in the test context:

```
testConfigurations[uniqueId]?.forEach {
            it.operation(forge)
        }
```

2. Then code will try to do `parent.get().callTestConfigurations(forge, operation)`, and in this case we will get executions, because test configurations are registered in the test class context.
3. And it will stop there (won't go up), because `parent.isPresent` will return `false` (since there is no parent for the top-most test class).

Do I get it right? And this parent chain can be more than 2 levels, because JUnit 5 allows nested test classes?

Reply:
Yes you're totally right, that's the way it work. And JUnit5 can have an unknown number of nested context (because of some test runners, eg parameterized tests, …)